ddtrace/internal/datadog/profiling/dd_wrapper/src/synchronized_sample_pool.cpp

```diff
@@ -0,0 +1,93 @@
+#include "synchronized_sample_pool.hpp"
+
+#include "libdatadog_helpers.hpp"
+
+extern "C"
+{
+#include "datadog/common.h"
+}
+
+namespace Datadog {
+
+void
+sample_delete_fn(void* sample)
+{
+    delete static_cast<Sample*>(sample);
+}
+
+SynchronizedSamplePool::SynchronizedSamplePool()
+{
+    size_t capacity = 4;
+    const char* capacity_env = std::getenv("DD_PROFILING_SAMPLE_POOL_CAPACITY");
+    if (capacity_env) {
+        try {
+            capacity = std::stoul(capacity_env);
+        } catch (const std::exception& e) {
+            std::cerr << "Failed to parse DD_PROFILING_SAMPLE_POOL_CAPACITY: " << e.what() << std::endl;
+        }
+        if (capacity == 0) {
+            std::cerr << "Invalid DD_PROFILING_SAMPLE_POOL_CAPACITY: " << capacity_env << std::endl;
+            capacity = 4;
+        }
+    }
+
+    ddog_ArrayQueue_NewResult array_queue_new_result = ddog_ArrayQueue_new(capacity, sample_delete_fn);
+    if (array_queue_new_result.tag == DDOG_ARRAY_QUEUE_NEW_RESULT_OK) {
+        pool = std::unique_ptr<ddog_ArrayQueue, Deleter>(array_queue_new_result.ok);
+    } else {
+        auto err = array_queue_new_result.err;
+        std::string errmsg = err_to_msg(&err, "Failed to create sample pool");
+        std::cerr << errmsg << std::endl;
+        ddog_Error_drop(&err);
+        pool = nullptr;
+    }
+}
+
+std::optional<Sample*>
+SynchronizedSamplePool::get_sample()
+{
+    if (pool == nullptr) {
+        // It's actually ok to call ddog_ArrayQueue_* methods with a nullptr,
+        // they will return an error result, but we already have printed out
+        // an error message in the constructor, so check for nullptr here to
+        // avoid spamming the error message.
+        return std::nullopt;
+    }
+    ddog_ArrayQueue_PopResult pop_result = ddog_ArrayQueue_pop(pool.get());
+    if (pop_result.tag == DDOG_ARRAY_QUEUE_POP_RESULT_OK) {
+        return static_cast<Sample*>(pop_result.ok);
+    } else if (pop_result.tag == DDOG_ARRAY_QUEUE_POP_RESULT_EMPTY) {
+        return std::nullopt;
+    } else if (pop_result.tag == DDOG_ARRAY_QUEUE_POP_RESULT_ERR) {
+        auto err = pop_result.err;
+        std::string errmsg = err_to_msg(&err, "Failed to get sample from pool");
+        std::cerr << errmsg << std::endl;
+        ddog_Error_drop(&err);
+        return std::nullopt;
+    }
+    // To silence the warning about missing return statement
+    return std::nullopt;
+}
```

Comment:
Here's an idea for not returning within conditionals
```suggestion
std::optional<Sample*>
SynchronizedSamplePool::get_sample()
{
    std::optional<Sample*> result = std::nullopt;

    // It's actually ok to call ddog_ArrayQueue_* methods with a nullptr,
    // they will return an error result, but we already have printed out
    // an error message in the constructor, so check for nullptr here to
    // avoid spamming the error message.
    if (pool != nullptr) {
        ddog_ArrayQueue_PopResult pop_result = ddog_ArrayQueue_pop(pool.get());

        if (pop_result.tag == DDOG_ARRAY_QUEUE_POP_RESULT_OK) {
            result = static_cast<Sample*>(pop_result.ok);
        } else if (pop_result.tag == DDOG_ARRAY_QUEUE_POP_RESULT_ERR) {
            auto err = pop_result.err;
            std::string errmsg = err_to_msg(&err, "Failed to get sample from pool");
            std::cerr << errmsg << std::endl;
            ddog_Error_drop(&err);
        }
    }

    return result;
}
```

Reply:
I was thinking of Rust's `match(pop_result.tag)` and being extra pedantic to list out all cases, but this looks better. Thanks for the suggestion.

Reply:
Yeah, the implicit return in rust is 👨‍🍳 💋  for this type of thing